source/grid/reference_cell.cc

```diff
@@ -807,6 +807,256 @@ ReferenceCell::gmsh_element_type() const
+        for (unsigned int i = 0; i < 3; ++i)
+          is_convex_combination = is_convex_combination &&
+                                  (0.0 <= combination_coordinates[i]) &&
+                                  (combination_coordinates[i] <= 1.0);
+        in_quad = is_convex_combination;
+      }
+    else
+      in_quad = 0.0 <= c0 && c0 <= 1.0 && 0.0 <= c1 && c1 <= 1.0;
+
+    if (in_quad)
+      return shifted_p.distance_square(projected_shifted_p);
+    else
+      return std::numeric_limits<double>::max();
+  }
+} // namespace
+
+
+
+template <int dim>
+double
+ReferenceCell::distance_to_unit_cell_square(const Point<dim> &p) const
+{
+  AssertDimension(dim, get_dimension());
+
+  // Handle simple cases first:
+  Assert(dim != 0, ExcNotImplemented());
+  if (contains_point(p, 0.0))
+    return 0.0;
+  if (dim == 1)
+    return distance_to_line_square(vertex<dim>(0), vertex<dim>(1), p);
+
+  // Find the closest vertex so that we only need to check adjacent faces and
+  // lines.
+  unsigned int closest_vertex_no        = 0;
+  double closest_vertex_distance_square = vertex<dim>(0).distance_square(p);
+  for (unsigned int i = 1; i < n_vertices(); ++i)
+    {
+      const double new_vertex_distance_square =
+        vertex<dim>(i).distance_square(p);
+      if (new_vertex_distance_square < closest_vertex_distance_square)
+        {
+          closest_vertex_no              = i;
+          closest_vertex_distance_square = new_vertex_distance_square;
+        }
+    }
+
+  double min_distance_square = std::numeric_limits<double>::max();
+  if (dim == 2)
+    {
+      for (const unsigned int face_no :
+           faces_for_given_vertex(closest_vertex_no))
+        {
+          const Point<dim> v0 = vertex<dim>(line_to_cell_vertices(face_no, 0));
+          const Point<dim> v1 = vertex<dim>(line_to_cell_vertices(face_no, 1));
+
+          min_distance_square =
+            std::min(min_distance_square, distance_to_line_square(v0, v1, p));
+        }
+    }
+  else
+    {
+      // If we can project via a face's normal vector then that's the shortest
+      // distance to the unit cell.
+      for (const unsigned int face_no :
+           faces_for_given_vertex(closest_vertex_no))
+        {
+          auto face_cell = face_reference_cell(face_no);
+          // We only need the first three points since for quads the last point
+          // is redundant
+          std::array<Point<dim>, 3> vertices;
+          for (unsigned int vertex_no = 0; vertex_no < 3; ++vertex_no)
+            vertices[vertex_no] = vertex<dim>(face_to_cell_vertices(
+              face_no, vertex_no, default_combined_face_orientation()));
+
+          min_distance_square = std::min(
+            min_distance_square,
+            distance_to_normal_quad_projection_square(vertices, p, face_cell));
+        }
+
+      // No normal projection was possible, so find the nearest point on a line
+      if (min_distance_square == std::numeric_limits<double>::max())
```

Comment:
3b. For the sake of completeness: the closest point in 3D may be on a line. In that case we can't do a normal projection.